tests/callbacks/test_swa.py

```diff
@@ -0,0 +1,134 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+
+import pytest
+import torch
+from torch import nn
+from torch.utils.data import DataLoader
+
+from pytorch_lightning import Trainer
+from pytorch_lightning.utilities import _PYTORCH_GREATER_EQUAL_1_6_0
+from tests.base import BoringModel, RandomDataset
+
+if _PYTORCH_GREATER_EQUAL_1_6_0:
+    from pytorch_lightning.callbacks import StochasticWeightAveraging
+
+    class SwaTestModel(BoringModel):
+
+        def __init__(self, batchnorm: bool = True):
+            super().__init__()
+            layers = [nn.Linear(32, 32)]
+            if batchnorm:
+                layers.append(nn.BatchNorm1d(32))
+            layers += [nn.ReLU(), nn.Linear(32, 2)]
+            self.layer = nn.Sequential(*layers)
+
+        def training_step(self, batch, batch_idx):
+            output = self.forward(batch)
+            loss = self.loss(batch, output)
+            return {"loss": loss}
+
+        def train_dataloader(self):
+            return DataLoader(RandomDataset(32, 64), batch_size=2)
+
+    class SwaTestCallback(StochasticWeightAveraging):
+        update_parameters_calls: int = 0
+        transfer_weights_calls: int = 0
+
+        def update_parameters(self, *args, **kwargs):
+            self.update_parameters_calls += 1
+            return StochasticWeightAveraging.update_parameters(*args, **kwargs)
+
+        def transfer_weights(self, *args, **kwargs):
+            self.transfer_weights_calls += 1
+            return StochasticWeightAveraging.transfer_weights(*args, **kwargs)
+
+        def on_train_epoch_start(self, trainer, *args):
+            super().on_train_epoch_start(trainer, *args)
+            assert trainer.train_loop._skip_backward == (trainer.current_epoch > self.swa_end)
+
+        def on_train_epoch_end(self, trainer, *_):
+            super().on_train_epoch_end(trainer, *_)
+            if self.swa_start <= trainer.current_epoch <= self.swa_end:
+                swa_epoch = trainer.current_epoch - self.swa_start
+                assert self.n_averaged == swa_epoch + 1
+            elif trainer.current_epoch > self.swa_end:
+                assert self.n_averaged == self._max_epochs - self.swa_start
+
+
+def train_with_swa(tmpdir, accelerator=None, gpus=None, num_processes=None):
+    model = SwaTestModel()
+    swa_callback = SwaTestCallback(swa_epoch_start=2, swa_lrs=0.005)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=4,
+        limit_train_batches=4,
+        callbacks=[swa_callback],
+        accelerator=accelerator,
+        gpus=gpus,
+        num_processes=num_processes
+    )
```

Comment:
Can you assert accumulate_grad_batches, etc.. having been reset properly ?

Reply:
That's already checked in the bottom test. Are you saying check it also for the accelerator tests?

Reply:
Did it for all tests